internal/attestation/azure/snp/validator.go

```diff
@@ -36,249 +41,372 @@ type Validator struct {
 	}
 	v.Validator = vtpm.NewValidator(
 		cfg.Measurements,
 		v.getTrustedKey,
-		validateCVM,
+		// stub, since SEV-SNP attestation is already verified in trustedKeyFromSNP().
+		func(vtpm.AttestationDocument, *attest.MachineState) error {
+			return nil
+		},
 		log,
 	)
 	return v
 }
 
-// validateCVM is a stub, since SEV-SNP attestation is already verified in trustedKeyFromSNP().
-func validateCVM(vtpm.AttestationDocument, *attest.MachineState) error {
-	return nil
-}
-
-func newSNPReportFromBytes(reportRaw []byte) (snpAttestationReport, error) {
-	var report snpAttestationReport
-	if err := binary.Read(bytes.NewReader(reportRaw), binary.LittleEndian, &report); err != nil {
-		return snpAttestationReport{}, fmt.Errorf("reading attestation report: %w", err)
-	}
-
-	return report, nil
-}
-
-func reverseEndian(b []byte) {
-	for i := 0; i < len(b)/2; i++ {
-		b[i], b[len(b)-i-1] = b[len(b)-i-1], b[i]
-	}
-}
-
 // getTrustedKey establishes trust in the given public key.
-// It does so by verifying the SNP attestation statement in instanceInfo.
+// It does so by verifying the SNP attestation document.
 func (v *Validator) getTrustedKey(ctx context.Context, attDoc vtpm.AttestationDocument, extraData []byte) (crypto.PublicKey, error) {
+	// transform the instanceInfo received from Microsoft into a verifiable attestation report format.
 	var instanceInfo azureInstanceInfo
 	if err := json.Unmarshal(attDoc.InstanceInfo, &instanceInfo); err != nil {
-		return nil, fmt.Errorf("unmarshalling instanceInfoRaw: %w", err)
+		return nil, fmt.Errorf("unmarshalling instanceInfo: %w", err)
 	}
-
-	report, err := newSNPReportFromBytes(instanceInfo.AttestationReport)
+	att, err := instanceInfo.attestationWithCerts(v.log, v.getter)
 	if err != nil {
 		return nil, fmt.Errorf("parsing attestation report: %w", err)
 	}
 
-	vcek, err := v.validateVCEK(instanceInfo.Vcek, instanceInfo.CertChain)
+	// Verify the attestation report's certificates.
+	trustedArk := x509.Certificate(v.config.AMDRootKey)             // ARK, specified in Constellation config.
+	ask, err := x509.ParseCertificate(att.CertificateChain.AskCert) // ASK, as reported from THIM / KDS.
 	if err != nil {
-		return nil, fmt.Errorf("validating VCEK: %w", err)
-	}
-
-	if err := v.validateSNPReport(ctx, vcek, report, instanceInfo.MAAToken, extraData); err != nil {
-		return nil, fmt.Errorf("validating SNP report: %w", err)
-	}
-
+		return nil, fmt.Errorf("parsing ASK certificate: %w", err)
+	}
+	// We verify the ASK certificate against the ARK certificate, since the ARK certificate is the root certificate
+	// and inherently trusted. This check should also be performed by go-sev-guest, but as we have to specify both ASK and ARK for
+	// the trusted roots, this needs to be done here as well.
+	rootPool := x509.NewCertPool()
+	rootPool.AddCert(&trustedArk)
+	if _, err := ask.Verify(x509.VerifyOptions{
+		Roots: rootPool,
+	}); err != nil {
+		return nil, fmt.Errorf("verifying ASK certificate: %w", err)
+	}
```

Comment:
```suggestion
```

Reply:
Thanks for pointing this out. I didn't catch the skipped verifier in the test. With the default verifier, this check is done in the library as expected.